New API for FixedPointingInfo, decouple from GADF metadata

maxnoe opened the pull request:
<!-- These are hidden commments. You can leave them or delete them. -->

<!-- Don't be shy! Github issues or pull requests are welcome any time. -->
<!-- But if you first want to chat, use the Gammapy Slack. -->
<!-- See https://gammapy.org/contact.html -->

**Description**
<!-- What is the motivation for this pull request? -->
<!-- Briefly: what changes are done here? -->
<!-- If this is to address a Github issue, mention its number to create a link -->
<!-- Example: This PR is to fix issue #42 -->

This pull request is here to further work on #3767 and to move towards a API of the pointing information that is easier to reason about.

The properties / methods that are not directly related to pointing are deprecated (and should be accessed directly from `Observation` or the `EventList` etc.

**Dear reviewer**
<!-- Let the reviewer and Gammapy team know what you want: -->
<!-- * Is this ready for review? Or is it work in progress and you want some feedback? -->
<!-- * Do you want to go through review here? Or if someone just finish this up and merge it in? -->
<!-- Do you have any specific questions, e.g. about API or implementation? -->
<!-- Do you include a test executing new code you're adding (to make sure it runs)? -->
<!-- Do you include some documentation? Is it needed? -->

adonath commented:
Thanks a lot @maxnoe! I haven't looked into any details here (can do this later...). However in general I was wondering whether you have thought about using on a `pydantic` model for defining the meta data model. It will solve the validation and type declarations...

maxnoe commented:
@registerrier @adonath 

The tests are passing now in principle. The errors are due to the deprecation warnings of all that I have deprecated and are mainly occuring in the test setup code, not gammapy itself.

Before I go start fixing the 90+ deprecation warnings this is raising now, could you please review the API changes and say if you agree with them?


Main changes are:

* the new API of FixedPointingInfo (if you ignore the deprecated parts) is essentially:
```
class FixedPointingInfo:
    mode: PointingMode
    fixed_icrs : None | SkyCoord[ICRS]
    fixed_altaz : None | SkyCoord[AltAz]

    def get_icrs(self, time, location) -> SkyCoord[AltAz]
    def get_altz(self, time, location) -> SkyCoord[AltAz]
    def from_fits_header(cls, format, version) -> FixedPointingInfo
    def to_fits_header(self, format, versoin) -> fits.Header
    def read(cls, path) -> FixedPointingInfo
```

All other parts are deprecated as they are either suboptimal (don't force giving a valid location / obstime) or have nothing to do with pointing (tstart, tstop, obstime, location).


I added pointing as a data member to `Observation` and reading it using a `LazyFits` using the event header.

I also added `get_pointing_icrs` and `get_pointing_altaz` to `Observation`, which are very quick shorthands, just taking the `location` from observation and by default the `mid of run time`. This is the current use in gammapy, but there are several open issues connected to this.

The other pointing related properties of Observation are now also deprecated to make it clear that one should access `Observation.pointing` or `Observation.get_pointing_{altaz, icrs}`.

Questions I have:

* Do we want to  fallback to reading pointing info from IRF HDUs if there is no events HDU? Or is that a case where users have to manually supply a pointing?
* Do we want to require the `time` argument in `Observation.get_pointing_{altaz,icrs}` explicitly? This would make it clear that a time is required and that using the "mid of observation time" really is a (over-) simplification.

registerrier reviewed (commented):
Thanks @maxnoe. We just discussed the new API and it looks fine.

maxnoe commented:
@adonath @registerrier Tests should be passing now

bkhelifi commented on gammapy/data/observations.py:
```diff
@@ -491,7 +542,9 @@ def write(self, path, overwrite=False, format="gadf", include_irfs=True):
         events = self.events
         if events is not None:
-            hdul.append(events.to_table_hdu(format=format))
+            events_hdu = events.to_table_hdu(format=format)
```
I have a question that might be related to an other PR.
As for each 'format' the versions can evolve, should be 'extended' the parameters list of the `to_table_hdu` function with a version ID?
The version of the format is written into the DL3 meta data and this information should be kept into the internal meta data such that it can be re-used.

maxnoe replied:
Yes, we should

maxnoe replied:
Relevent issue #4239

adonath commented on gammapy/data/observations.py:
```diff
@@ -286,29 +308,56 @@ def obs_info(self):
 
     @property
+    @deprecated("v1.1", message="Use observation.get_pointing_icrs(time) instead")
     def pointing_radec(self):
```
For convenience one could  still introduce `obs.pointing_icrs_tmid`. Otherwise many people will end up writing `obs.get_pointing_icrs(obs.tmid)` it is not that bad, but the former is still better readable...

maxnoe replied:
Actually, my reasoning was to not make it convenient to assume the pointing doesn't change. People should get the "oh right, it might be time-dependent" when trying to get the pointing.

adonath commented on gammapy/data/pointing.py:
```diff
@@ -67,45 +70,194 @@ class FixedPointingInfo:
+            raise ValueError("mode=DRIFT is only supported by GADF 0.3")
+
+        header = fits.Header()
+        if self.mode == PointingMode.POINTING:
```
Could move this code to the `PointingMode` class directly...would lead to a shorter implementation here...

maxnoe replied:
What could be moved? I don't see what part of the code is dealing with the `PonitingMode` itself, the `PointingMode` just decides which information stored in `PointingInfo` is put into the header.

adonath replied:
Oh yes, you are right. As we already have `PointingMode.from_gadf_string()` I thought it might make sense to have `PointingMode.to_fits_header()` as well, but it would contain effectively only two lines of code and the returned header would be barely useful. So fine as is...

adonath commented on gammapy/data/pointing.py:
```diff
@@ -67,45 +70,194 @@ class FixedPointingInfo:
+            legacy_altaz=legacy_altaz,
+        )
+
+    def to_fits_header(self, format="gadf", version="0.3", time_ref=None):
```
Same here: if the method is public please add a docstring...

maxnoe replied:
Same? Is there another one missing? 

I added one for this method now.

maxnoe replied:
Ah, from_fits_header, fixing now

adonath reviewed (commented):
Thanks @maxnoe, I have left some more minor comments...

maxnoe commented:
Ok, this should be it: comments addressed and rebased

registerrier reviewed (approved):
Thanks @maxnoe . No further comment from my side.
